File: pages/refactorings/replace-mutable-fields-with-lenses/03-file-tree/After.hs
-- Same, with a lens on the size field: the write goes through the
-- lens; the recursion over children is unchanged.
module After where

data Node = Node { name :: String, size :: Int, children :: [Node] }
  deriving (Eq, Show)

data Lens s a = Lens { get :: s -> a, set :: a -> s -> s }

modify :: Lens s a -> (a -> a) -> s -> s
modify l f s = set l (f (get l s)) s

sizeLens :: Lens Node Int
sizeLens = Lens { get = size, set = \v n -> n { size = v } }

bumpSizes :: Node -> Int -> Node
bumpSizes n by =
  modify sizeLens (+ by)
    (n { children = map (`bumpSizes` by) (children n) })

File: pages/refactorings/replace-mutable-fields-with-lenses/03-file-tree/After.scala
// Same, with a lens on the size field: the write goes through the
// lens; the recursion over children is unchanged.
object After:
  case class Node(name: String, size: Int, children: List[Node])

  case class Lens[S, A](get: S => A, set: A => S => S):
    def modify(f: A => A): S => S = s => set(f(get(s)))(s)

  val size: Lens[Node, Int] =
    Lens[Node, Int](_.size, v => n => n.copy(size = v))

  def bumpSizes(n: Node, by: Int): Node =
    size.modify(_ + by)(
      n.copy(children = n.children.map(bumpSizes(_, by))))
